ENH: Add sheet name to error message when reading Excel

aaossa opened the pull request:
- [X] Closes #48706
- [X] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] ~Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.~
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

---

**Description of changes**

When reading Excel files with `BaseExcelReader`, is possible that `TextParser` raises an `Exception` on different situations. As the reader works on a sheet basis, including the name of the sheet that caused the exception can be useful for debugging.

The implemented solution captures any `Exception` that gets raised while parsing and reading a sheet from an Excel file, then adds the name of the sheet to the exception message and "re-raises" the exception.

aaossa committed:
Add sheet to error message when reading Excel

When reading Excel files with `BaseExcelReader`, is possible that
`TextParser` raises `ValueError` on different situations. As the
reader works on a sheet basis, including the name of the sheet that
caused the exception can be useful for debugging.

The implemented solution captures any `ValueError` that gets raised
while parsing and reading a sheet from an Excel file, then adds the
name of the sheet to the exception message and "re-raises" the
exception using the `Exception.with_traceback` method.

Why use `with_traceback`? This allow-us to maintin the focus on the
actual line that raised the ValueError instead of directing the
attention to the introduced block of code.

Signed-off-by: Antonio Ossa Guerra <aaossa@uc.cl>

aaossa committed:
Test sheet name presence in `ValueError` message

The test performs a `read_excel` operation on a blank Excel file while
also passing `[1]` as header value. This operations results in a
`ValueError`.

Before the fix, the error message did not display any information
about the offending sheet, making it harder to debug Excel files with
several sheets. For this reason, the value of `sheet_name` was
explicitly set to its default value (`None`).

The test compares the `ValueError` message with the expected message
that contains the offending sheet name.

Signed-off-by: Antonio Ossa Guerra <aaossa@uc.cl>

aaossa committed:
Add whatsnew not on improved error message

The introduced enhancement of the error message on `read_excel` is now
included in the whatsnew

Signed-off-by: Antonio Ossa Guerra <aaossa@uc.cl>

aaossa committed:
Merge branch 'main' into issue-48706

aaossa commented on pandas/io/excel/_base.py:
```diff
@@ -901,6 +901,10 @@ def parse(
                 # No Data, return an empty DataFrame
                 output[asheetname] = DataFrame()
 
+            except ValueError as err:
```
**TLDR**: Arguments that accept callable can propagate **any exception** from those callables. Changing `except ValueError as err` to `except Exception as err` adds the sheet name to any case.

---

To answer your question I was exploring other places that end up in the added except. The sheet name is added when `read_excel` raises an exception while parsing an Excel file. I modified the source code to see which tests fail with an unexpected exception related to the parse operation (raised `SyntaxError` for this purpose). The cited test do not fail at the moment because they test internals or because the regex does not match the string completely, but they still match (because it was added at the end).

These are some of the cases where the sheet name is included in the error message (if the PR changes are applied):

* `tests/io/excel/test_readers.py`
    * `ValueError` when passing non existent columns (`test_usecols_pass_non_existent_column`)
    * `ValueError` when passing wrong type element on `usecols` argument (`test_usecols_wrong_type`)
    * `ValueError` when trying to convert column to an incompatible dtype (`test_reader_dtype`)
* `tests/io/excel/test_writers.py`
    * `ValueError` when trying to use `mangle_dupe_cols=False` (`test_duplicated_columns`)
* `tests/io/parser/test_python_parser_only.py`
    * `ValueError` when passing invalid `skipfooter` (non int) (`test_invalid_skipfooter_non_int`)
    * `ValueError` when passing invalid `skipfooter` (negative) (`test_invalid_skipfooter_negative`)
    * `ValueError` when there is a bad row while parsing and using `skipfooter` (`test_skipfooter_bad_row`)
    * `ValueError` when there's a malformed `skipfooter` (`test_malformed_skipfooter`)

I was able to add (almost) all of these cases in the test, but I'm not sure if this is required:

```diff
diff --git a/pandas/tests/io/excel/test_readers.py b/pandas/tests/io/excel/test_readers.py
index c725325086..d07c660dfb 100644
--- a/pandas/tests/io/excel/test_readers.py
+++ b/pandas/tests/io/excel/test_readers.py
@@ -657,11 +657,23 @@ class TestReaders:

     def test_value_error_message_includes_sheet_name(self, read_ext):
         # GH 48706
-        msg = (
-            r"Passed header=\[1\], len of 1, but only 1 lines in file \(sheet: Sheet1\)"
-        )
-        with pytest.raises(ValueError, match=msg):
+        with pytest.raises(ValueError, match=r" \(sheet: Sheet1\)$"):
             pd.read_excel("blank_with_header" + read_ext, header=[1], sheet_name=None)
+        with pytest.raises(ValueError, match=r" \(sheet: 0\)$"):
+            pd.read_excel("test1" + read_ext, usecols=["E"])
+        with pytest.raises(ValueError, match=r" \(sheet: 0\)$"):
+            pd.read_excel("test1" + read_ext, usecols=["E1", 0])
+        with pytest.raises(ValueError, match=r" \(sheet: 0\)$"):
+            pd.read_excel("testdtype" + read_ext, dtype={"d": "int64"})
+        with pytest.raises(ValueError, match=r" \(sheet: 0\)$"):
+            pd.read_excel("test1" + read_ext, skipfooter="non int")
+        with pytest.raises(ValueError, match=r" \(sheet: 0\)$"):
+            pd.read_excel("test1" + read_ext, skipfooter=-1)
+        with pytest.raises(ValueError, match=r" \(sheet: 0\)$"):
+            pd.read_excel("test1" + read_ext, header=None, mangle_dupe_cols=False)
+        # TODO: test_skipfooter_bad_row
+        # TODO: test_malformed_skipfooter
+        # TODO: test_on_bad_lines_callable_dont_swallow_errors

     @pytest.mark.filterwarnings("ignore:Cell A4 is marked:UserWarning:openpyxl")
     def test_date_conversion_overflow(self, request, engine, read_ext):

```

That said, **arguments that accept a callable can cause any kind of exception**. See these examples:

```python
with pytest.raises(ValueError, match=r" \(sheet: 0\)$"):
    pd.read_excel("test1" + read_ext, usecols=lambda x: 1/0)
with pytest.raises(ValueError, match=r" \(sheet: 0\)$"):
    pd.read_excel("test1" + read_ext, skiprow=lambda x: 1/0)
```

These test fail because the exception is not `ValueError` but `ZeroDivisionError`. If we change the exception, then the match fails because it does not include the sheet name at the moment.

rhshadrach replied:
+1 on setting sheetname for any Exception. While it could be superfluous in some cases, it is certainly not harmful and can be helpful in other cases.

rhshadrach commented on pandas/io/excel/_base.py:
```diff
@@ -901,6 +901,10 @@ def parse(
 
+            except ValueError as err:
+                err.args = (f"{err.args[0]} (sheet: {asheetname})", *err.args[1:])
+                raise err.with_traceback(err.__traceback__)
```
Does this just set `err.__traceback__` to be `err.__traceback__`, or is it doing something else as well? E.g.

```
def foo(x):
    raise ValueError('this is the original')

try:
    foo(5)
except ValueError as err:
    raise err
```

Both this and replacing the last line with `raise err.with_traceback(err.__traceback__)` give the same output:

```
Traceback (most recent call last):
  File ".../scratch.py", line 26, in <module>
    raise err
  File ".../scratch.py", line 24, in <module>
    foo(5)
  File ".../scratch.py", line 21, in foo
    raise ValueError('this is the original')
ValueError: this is the original
```

aaossa committed:
Merge branch 'main' into issue-48706

aaossa committed:
Use simpler syntax for exception chaining

Instead of handling traceback of the capture exception manually, this
simpler syntax looks better, is clearer, and produces the same result

Signed-off-by: Antonio Ossa Guerra <aaossa@uc.cl>

aaossa committed:
Add sheet name to any Exception traceback

Instead of including the sheet name on the traceback when a ValueError
is raised, we'll do it for any kind of Exception. In particular, this
is because some of the arguments for `read_excel` are callables, and
they can raise any Exception. Including the sheet name might be
helpful in some of those cases, so we'll capture them too

Signed-off-by: Antonio Ossa Guerra <aaossa@uc.cl>

aaossa committed:
Update whatsnew about enhanced error message

The previous message only included `ValueError` as the messages that
displayed the sheet name on their traceback, but that behavior is now
present on any exception

Signed-off-by: Antonio Ossa Guerra <aaossa@uc.cl>

aaossa committed:
Merge branch 'main' into issue-48706

mroeschke reviewed (approved):
LGTM. Merge when ready @phofl or @rhshadrach

phofl commented:
thx @aaossa
